app/domain/authentication/resource_restrictions/fetch_non_empty_resource_annotations.rb

```diff
@@ -0,0 +1,46 @@
+require 'command_class'
+
+module Authentication
+  module ResourceRestrictions
+
+    FetchNonEmptyResourceAnnotations = CommandClass.new(
+      dependencies: {
+        role_class: ::Role,
+        resource_class: ::Resource,
+      },
+      inputs: %i[account role_name]
+    ) do
+
+      def call
+        fetch_resource_annotations
+      end
+
+      private
+
+      def fetch_resource_annotations
+        resource_annotations
+      end
+
+      def resource_annotations
+        resource.annotations.each_with_object({}) do |annotation, result|
+          annotation_values = annotation.values
+          value = annotation_values[:value]
+          raise Errors::Authentication::ResourceRestrictions::EmptyAnnotationGiven, annotation_values[:name] if value.blank?
+
+          result[annotation_values[:name]] = value
+        end
+      end
+
+      def resource
+        # Validate role exists, otherwise getting role annotations return empty hash.
+        role_id = @role_class.roleid_from_username(@account, @role_name)
+        resource = @resource_class[role_id]
+
+        raise Errors::Authentication::Security::RoleNotFound, role_id unless resource
+
+        resource
+      end
```

Comment:
```suggestion
      def resource
        return @resource if @resource
        
        @resource = @resource_class[role_id]
        raise Errors::Authentication::Security::RoleNotFound, role_id unless @resource
        @resource
      end
      
      def role_id
        @role_id ||= @role_class.roleid_from_username(@account, @role_name)
      end
```

Reply:
@orenbm What's the meaning of making `@resource` a class member?

Reply:
it's just for memoization. so you don't run the whole logic every time you call the `resource` method.